File: scalatest/src/Trie.scala
/**
 * Created with IntelliJ IDEA.
 * User: mactep
 * Date: 04.10.13
 * Time: 14:18
 */
class Trie {
  private var _trie : Impl = new Impl()
  private val _cont : Cont = new Cont(_trie.root())

  def copyof(trie : Trie) = {
    _trie = trie._trie
    _cont.copyof(trie._cont)
  }

  def insert(i : Int, symbol : Byte) : Int = {
    val node = _trie.insert(_cont.nodeof(i), symbol)
    if (node.id() == _cont.size()) {
      _cont.push(node)
    }
    return node.id()
  }

  def symbolof(i : Int) : Int = {
    return _cont.nodeof(i).symbol()
  }

  def setdataof(i : Int, data : Any) = {
    _cont.setdata(i, data)
  }

  def dataof(i : Int) : Any = {
    return _cont.dataof(i)
  }

  def parentof(i : Int) : Int = {
    val node = _cont.nodeof(i).parent()
    if (node != null) {
      return node.id()
    }
    return -1
  }

  def nextof(i : Int, symbol : Byte) : Int = {
    return _cont.nodeof(i).get(symbol).id()
  }

  def keysof(i : Int) : Iterable[Byte] = {
    return _cont.nodeof(i).keys()
  }

  def isfork(i : Int) : Boolean = {
    return _cont.nodeof(i).size() > 1
  }

  def isleaf(i : Int) : Boolean = {
    return _cont.nodeof(i).size() == 0
  }

  def size() : Int = {
    return _trie.size()
  }
}

File: trietest/src/Trie.java
import java.util.Iterator;
import java.util.Vector;

/**
 * Created with IntelliJ IDEA.
 * User: mactep
 * Date: 04.10.13
 * Time: 10:50
 */
public class Trie<D> implements Iterable<Integer> {
    private TrieImplementation implementation = new TrieImplementation();
    private TrieContainer<D>   container      = new TrieContainer<D>();
    private Vector<Integer>    dfs_cache      = new Vector<Integer>();

    public Trie() {
        container.push(implementation.getRoot());
    }

    public <C> void copyOf(Trie<C> trie) {
        implementation = trie.implementation;
        container.copyOf(trie.container);
        dfs_cache = trie.dfs_cache;
    }

    public int insert(int i, byte symbol) {
        TrieNode node = implementation.insert(container.nodeAt(i), symbol);
        if(node.getId() == container.size()) {
            container.push(node);
        }
        return node.getId();
    }

    public byte symbolOf(int i) {
        return container.nodeAt(i).getSymbol();
    }

    public void setDataOf(int i, D data) {
        container.setDataAt(i, data);
    }

    public D dataOf(int i) {
        return container.dataAt(i);
    }

    public int parentOf(int i) {
        TrieNode node = container.nodeAt(i).getParent();
        if(node != null) {
            return node.getId();
        }

        return -1;
    }

    public int nextOf(int i, byte symbol) {
        return container.nodeAt(i).get(symbol).getId();
    }

    public boolean isFork(int i) {
        return container.nodeAt(i).keys().size() > 1;
    }

    public boolean isLeaf(int i) {
        return container.nodeAt(i).keys().size() == 0;
    }

    public int size() {
        return implementation.getSize();
    }

    @Override
    public Iterator<Integer> iterator() {
        return null;
    }
}
